### PR TITLE
Logic for checking out and syncing from the mainline project repo

### PR DESCRIPTION
**Pull request by bion**
This PR introduces the concept of a `checkout` directory that contains a sparse git checkout of the mainline CiviForm project, containing only the `cloud/` subtree.

For any script in the deploy that needs access to cloud automation tooling (Terraform, cloud-specific scripts, etc), the script specifies the version it intends to interact with (defaulting to latest) and uses bash library functions to synchronize `checkout` to the corresponding git commit for the mainline repo.

This PR starts with the `bin/deploy` script, which doesn't actually deploy anything yet, but exercises this checkout logic.

In general we should seek to keep as much logic as we can in the mainline repo, strictly limiting this deploy repo and its forks to deployment-specific configuration along with the minimum amount of logic needed to access the scripts and generic IaC code in the mainline repo.

### CODE REVIEW

File: bin/lib/checkout.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  rm -rf checkout
+  mkdir checkout
+
+  pushd checkout > /dev/null
+
+  git init --initial-branch=main
+  git remote add origin http://github.com/seattle-uat/civiform
+  git config core.sparseCheckout true
+
+  echo '/cloud' >> .git/info/sparse-checkout
```
Comment (goldblatt) [inline bin/lib/checkout.sh:47]: nit: does this need single quotes or can we use double

File: bin/deploy [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#! /usr/bin/env bash
+
+# DOC: Deploy an image to an environment. Usage: deploy [--tag=imageTag] [env]; env defaults to "staging", tag defaults to "latest".
```
Comment (goldblatt) [inline bin/deploy:3]: i feel like the naming here is a lil confusing to me as it's not deploying anything is it?

**Review by goldblatt — CHANGES_REQUESTED** [review]

File: bin/lib/checkout.sh (hunk 1 above)
Comment (bion) [reply bin/lib/checkout.sh]: done

File: bin/deploy (hunk 2 above)
Comment (bion) [reply bin/deploy:3]: Yep, this script isn't fully implemented yet. The checkout system is meant to be transparent to the user so I just chose one of the commands we plan to implement to serve as a stub to get this first piece of functionality in.

Comment (goldblatt) [reply bin/deploy:3]: right right it does say that in the pr

File: bin/lib/snapshots.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#######################################
+# Predicate function that checks if a string is formatted
+# as a CiviForm server image snapshot tag.
+# Arguments:
+#   1: The string to test.
+#######################################
+function snapshots::tag_is_snapshot() {
+  local full_tag="${1}"
+
+  [[ "${full_tag}" =~ ^SNAPSHOT\-[a-zA-Z0-9]+\-[0-9]+$ ]]
```
Comment (goldblatt) [inline bin/lib/snapshots.sh:12]: nit: i think you can make this a variable

File: bin/deploy [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+done
+
+if [[ "${deploy_env}" != "staging" ]] && [[ "${deploy_env}" != "prod" ]]; then
+  out::error "Deployment env must be either prod or staging."
+  exit 1
+fi
+
+checkout::ensure_initialized
+
+if snapshots::tag_is_snapshot "${image_tag}"; then
```
Comment (goldblatt) [inline bin/deploy:30]: would say lil confused by the terminology here like tag, snapshot and sha all seem like the are kind of similar. image_tag also seems lil confusing to me. struggling to get my git terminology. would maybe just add some comments here?

File: bin/lib/snapshots.sh (hunk 3 above)
Comment (bion) [reply bin/lib/snapshots.sh:12]: the regex?

Comment (goldblatt) [reply bin/lib/snapshots.sh:12]: yeah

**Review by goldblatt — COMMENTED** [review]
approve with kind of caveat that i think some of the terminology could use a comment or two

**Review by goldblatt — APPROVED** [review]

File: bin/lib/snapshots.sh (hunk 3 above)
Comment (bion) [reply bin/lib/snapshots.sh:12]: done

File: bin/deploy (hunk 4 above)
Comment (bion) [reply bin/deploy]: Added some comments and factored out another (hopefully well-named) function. PTAL

File: bin/lib/checkout.sh [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+  popd > /dev/null
+  echo "done"
+}
+
+#######################################
+# Pulls the latest git revisions for the checkout directory from the mainline
+# CiviForm repo.
+#######################################
+function checkout::sync() {
```
Comment (goldblatt) [inline bin/lib/checkout.sh:107]: would maybe call this at_main() instead of sync?

File: bin/lib/snapshots.sh [hunk 6]
```diff
@@ -0,0 +1,38 @@
+#! /usr/bin/env bash
+
+# This file contains shell functions for working with snapshot tags.
+#
+# A snapshot tag is a docker image tag assigned to CiviForm images when they are
```
Comment (goldblatt) [inline bin/lib/snapshots.sh:5]: this was super super helpful thanks for adding this in

File: bin/deploy (hunk 4 above)
Comment (goldblatt) [reply bin/deploy]: A++++ super helpful

File: bin/lib/checkout.sh (hunk 5 above)
Comment (bion) [reply bin/lib/checkout.sh]: done

**Review by zaralouis-sf — APPROVED** [review]
Everything looks good to me!

File: bin/lib/snapshots.sh [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# are first pushed to Docker Hub. All commits in the main CiviForm branch get an
+# image tag of the form:
+#
+# 'SNAPSHOT-[git short commit SHA]-[timestamp in unix seconds]'
+#
+# See https://github.com/seattle-uat/civiform/blob/main/bin/build-and-push
+
+# Matches a CiviForm snapshot tag with a capture group for the commit short SHA.
+readonly CIVIFORM_SNAPSHOT_TAG_REGEX="^SNAPSHOT\-([a-zA-Z0-9]+)\-[0-9]+$"
```
Comment (shanemc-goog) [inline bin/lib/snapshots.sh:14]: [0-9] -> \d

File: bin/deploy [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#! /usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (bion) [reply bin/deploy:1]: executables should have no extension, library code should have .sh: https://google.github.io/styleguide/shellguide.html#s2.1-file-extensions

File: bin/lib/snapshots.sh (hunk 7 above)
Comment (bion) [reply bin/lib/snapshots.sh:14]: i could not get that to work, i do not know why

Comment (shanemc-goog) [reply bin/lib/snapshots.sh:14]: \\d maybe? but no worries

**Review by shanemc-goog — APPROVED** [review]

File: bin/lib/snapshots.sh (hunk 7 above)
Comment (bion) [reply bin/lib/snapshots.sh:14]: > The irony that I had to write that with 4 \ s :D
😆 

I tried it with a variety of different numbers of backslashes to no avail 🤷🏻

**Review by shubha-rajan — APPROVED** [review]

**Pull request merged**